### PR TITLE
Backup expirations

### PR DESCRIPTION
**Pull request by ahmetmircik**
fixes https://github.com/hazelcast/hazelcast/issues/9153

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
@jerrinot can you please have a look again?

File: hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java [hunk 1]
```diff
@@ -293,27 +347,95 @@ private boolean notHaveAnyExpirableRecord(PartitionContainer partitionContainer)
...
+        for (RecordStore recordStore : maps.values()) {
+            sendExpiredKeysToBackups(recordStore, false);
         }
     }
 
+    public void sendExpiredKeysToBackups(RecordStore recordStore, boolean checkIfReachedBatch) {
+        InvalidationQueue invalidationQueue = recordStore.getExpiredKeys();
+
+        int size = invalidationQueue.size();
+        if (size == 0 || checkIfReachedBatch && size < MAX_EXPIRED_KEY_COUNT_IN_BATCH) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java]: shouldn't this also by take time-from-last-sending into consideration? 

imagine a single key expires. they I presume you want to eventually send this information to the backup even when no other key expires after that. something like the Nagle's algorithm basically.

**Comment by jerrinot** [comment]
@ahmetmircik: did you consider having a (hidden?) switch to re-introduce to old behavior? perhaps there are situations where extra backup operations do more harm than good. the impl is there anyway due the Rolling Upgrade.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/EvictBatchBackupOperation.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public void run() {
+        assert getNodeEngine().getClusterService().getClusterVersion().isGreaterOrEqual(Versions.V3_9);
+
+        for (Data key : expiredKeys) {
+            recordStore.evict(key, true);
+        }
+
+        // equalize backup entry count to owner entry count to have identical memory occupancy
+        int diff = recordStore.size() - ownerPartitionEntryCount;
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/EvictBatchBackupOperation.java]: can you please elaborate this part a bit? why is it needed?

**Comment by jerrinot** [comment]
I am thinking about this scenario:
1. key `K` expires. the primary put `K` into the expiration queue (for batching)
2. someone does `map(K, foo)` and inserts the same key again. This will create a new entry on a primary  and overwrites the backup entry on the backup replica.
3. eventually the expired keys are sent from primary to the backup replica. this will remove the key `K` from the backup replica even when it should not.

is this scenario realistic? is there anything we can do against it?

File: hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java (hunk 1 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java]: It is already like that. Background expiration task sets `checkIfReachedBatch` to `false` and it scans queue periodically regardless of the batch size.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/EvictBatchBackupOperation.java (hunk 2 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/EvictBatchBackupOperation.java]: please see [this](https://hazelcast.atlassian.net/wiki/display/EN/Partition+Owner+Should+Be+Responsible+for+Expirations+and+Evictions+on+Replicas+-+Design?focusedCommentId=131339648#comment-131339648)

**Comment by ahmetmircik** [comment]
@jerrinot yes good catch. It is possible. I merged a fix for that here.

File: hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java [hunk 3]
```diff
@@ -161,6 +186,12 @@ private int getInteger(String propertyName, int defaultValue) {
         return property == null ? defaultValue : parseInt(property);
     }
 
+    private boolean getBoolean(String propertyName, boolean defaultValue) {
+        Config config = nodeEngine.getConfig();
+        String property = config.getProperty(propertyName);
+        return property == null ? defaultValue : Boolean.parseBoolean(property);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/eviction/ExpirationManager.java]: fixed by using `HazelcastProperty`instead

**Pull request merged**